### PR TITLE
Dragon drop poof of conscepter

### PR DESCRIPTION
**Pull request by dabreegster**
screencast
2e5460542d6384468cb7edb041a21ffc

Here's the ~1 hour take on reoderable drag-and-drop cards. The color transformations to show what we're hovering on and where we're considering dropping a card are definitely awful. But if this seems initially useful, then I can work this into the road editor for reordering lanes.

I guess the main design question is: do we need any sort of animation to make this really work?

### CODE REVIEW

File: widgetry_demo/src/lib.rs [hunk 1]
```diff
@@ -315,12 +315,17 @@ fn setup_scrollable_canvas(ctx: &mut EventCtx) -> Drawable {
...
+        .map(|i| (i, make_draggable_card(ctx, i)))
+        .collect::<Vec<_>>();
     let style = ctx.style();
 
     let mut tabs = TabController::new("demo_tabs");
 
     let gallery_bar_item = style.btn_tab.text("Component Gallery");
     let gallery_content = Widget::col(vec![
+        "Reorder the cards below".text_widget(ctx),
+        DragDrop::new_widget(ctx, draggable_cards),
```
Comment (michaelkirk) [inline widgetry_demo/src/lib.rs]: I intended the gallery to show off the most useful common widgets first - could you move this more esoteric one towards the bottom.

**Comment by dabreegster** [comment]
screencast
Here's the absolute minimal demo of this in the road editor. Obviously the lane card style got totally lost

**Review by michaelkirk — COMMENTED** [review]
🔥 100% for the pun 🔥 

regarding the ui - the biggest thing that seems off is that the card you're "dragging" should follow the mouse cursor as you drag. Without that feedback, it's not clear to me that I'm doing a "drag and drop" operation.

There should be some kind of treatment to show the floating/dragging state - ideally a conventional drop shadow styling, but since that might be hard, any kind of border or mask that implies depth might be enough.

Without that additional feedback to intuitively tell me that I'm dragging something, I think I prefer the simpler "button" based editing. It's probably slower, but it's easier to understand. 

I'm willing to be overruled though - curious to hear what others think!

Other things, though IMO less important than the former:

1. As you drag, creating a gap between the two cards that visually shows where the drop target is, so you can clearly see how ordering will be affected.
2. Animating the cards movement on mouse-down and mouse-up to it's floating vs. resting position.

File: widgetry_demo/src/lib.rs (hunk 1 above)
Comment (dabreegster) [reply widgetry_demo/src/lib.rs]: Will do! This was just convenient for rapid prototyping

**Comment by dabreegster** [comment]
screencast
Here's a bit more progress, but there are some pretty bad bugs with the math for swapping elements and for offsetting the width when dragging.

> the biggest thing that seems off is that the card you're "dragging" should follow the mouse cursor as you drag
> As you drag, creating a gap between the two cards that visually shows where the drop target is, so you can clearly see how ordering will be affected.

Agreed! Both of these are now on their way to working. I found https://jqueryui.com/sortable/ as a reference for behavior of how this ought to work; it's much easier to try to reproduce behavior than it is to also have to come up with the right behavior.

**Comment by michaelkirk** [comment]
> Here's a bit more progress, but there are some pretty bad bugs with the math for swapping elements and for offsetting the width when dragging.

There's more we could do with animation to transition the user between states, and a drop shadow would be cool, but I think even without those things, it's likely worth using this drag-and-drop interface for lane editing, provided the egregious offset bugs can be fixed. Nice progress!

**Comment by michaelkirk** [comment]
https://github.com/a-b-street/abstreet/compare/master...michaelkirk:mkirk/dragondrop?expand=1

I got inspired (trapped on a plane) and worked on some of these issues:
- Shift the adjacent cards to "insert" rather than swap
- dragging 1st item looked fine, but any others had a weird offset
- dragged item would change the size of the widget, causing other screen content to be shifted around as you dragged

(also rebased)

I'm going punt it back to you now @dabreegster

**Comment by michaelkirk** [comment]
> dragged item would change the size of the widget, causing other screen content to be shifted around as you dragged

interestingly, I didn't see this behavior in your screen capture! I wonder if you had another local commit?

**Comment by dabreegster** [comment]
Closing, #728 took over this and is merged

**Pull request closed without merging**